File: daily_problems/2025/05/0521/personal_submission/cf1080d_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/fenwicktree"
// #include "atcoder/lazysegtree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divided, I conquered!

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int t;
    cin >> t;

    vector<long long> f(32, 0);
    for (int i = 1; i < 32; i ++)
        f[i] = 4 * f[i - 1] + 1;
    
    while (t --) {
        long long n, k;
        cin >> n >> k;

        if (n > 31) cout << "YES " << n - 1 << '\n';
        else {
            int ans = -1;
            for (int i = 0; i < n; i ++) {
                int cuts = n - i;
                long long left = (1ll << (cuts + 1)) - 2 - cuts;
                long long right = f[n] - ((1ll << (cuts + 1)) - 1) * f[i];
                if (left <= k && k <= right) {
                    ans = i;
                    break;
                }
            }

            if (ans == -1) cout << "NO\n";
            else cout << "YES " << ans << '\n';
        }
    }

    return 0;
}

File: daily_problems/2025/05/0521/personal_submission/cf1080d_yawn_sean.py
# Submission link: https://codeforces.com/contest/1080/submission/320573711
def main():
    t = II()
    outs = []

    f = [0]

    for _ in range(31):
        f.append(4 * f[-1] + 1)

    for _ in range(t):
        n, k = MII()
        if n > 31: outs.append(f'YES {n - 1}')
        else:
            for i in range(n):
                cuts = n - i
                left = (1 << cuts + 1) - 2 - cuts
                right = f[n] - ((1 << cuts + 1) - 1) * f[i]
                if left <= k <= right:
                    outs.append(f'YES {i}')
                    break
            else:
                outs.append('NO')

    print('\n'.join(outs))